### DISCUSSION TITLE
Saving Polars DataFrame as xlsx with embedded images - analogous function to PandasTools SaveXlsxFromFrame

### DISCUSSION THREAD

**Discussion by mpm896** [rdkit/rdkit#7537] [Show and tell]
Hello,

I've been using `Polars` recently instead of `Pandas` and haven't found an `rdkit` function to save a `Polars DataFrames` as an Excel sheet with embedded `Mol` images. I modified `PandasTools SaveXlsxFromFrame` to work with `Polars DataFrame` objects. I hope this can be helpful! The function can be found [here in my repo](https://github.com/mpm896/HTSCluster/blob/main/HTSCluster/utils/polars_xlsx.py).

**Comment by bertiewooster** [comment]
Glad you're combining the RDKit and Polars! Nice tool; it worked for me. 

One note: On a Mac, I had to set the file extension to xlsx to get the file to open (in Numbers; I don't have Excel), so you might put a note to that effect in the docstring. You might even set a default value of e.g. "xlsx_from_polarsdf_out.xlsx" so the file will be saved in the current working directory.

### Avoiding storing Mol objects in Polars dataframe
While I appreciate that you're making an analogous tool to [`PandasTools SaveXlsxFromFrame`](https://rdkit.org/docs/source/rdkit.Chem.PandasTools.html#rdkit.Chem.PandasTools.SaveXlsxFromFrame), as I understand it the [best practice is to put Polars-native data types in a dataframe, rather than Python objects](https://discord.com/channels/908022250106667068/1208523092385599559/1208800918942908423) such as RDKit Mol. Here's an example of why: I had a data frame with ~3500 Mol objects in a [Jupyter Notebook](https://bertiewooster.github.io/2024/05/01/Tautomer-Sources-Comparison.html), and the Python kernel kept crashing. After I removed the objects from the df, the crashes stopped. (I replaced each Mol object with the properties I needed from it, for example its SMILES, and the SMILES of its tautomers.)

So perhaps it would be useful to allow `xlsx_from_polarsdf` to accept a Polars df that contains a SMILES string (or another identifier that can be made into a molecule), then the utility creates the molecule on the fly, and then creates the image. For example, at line 79:

```
sml = row[col_idx]
mol = Chem.MolFromSmiles(sml)
img = Draw.MolToImage(mol ...)
```
